tests/Midi_gtest.cpp

```diff
@@ -0,0 +1,288 @@
+        event = midi.PopEvent();
+    }
+
+    //help testing
+    void Test(uint8_t evType,
+              uint8_t chkType,
+              uint8_t chn,
+              bool    twoData,
+              uint8_t d0,
+              uint8_t d1)
+    {
+        //test eventtype == expected type
+        EXPECT_EQ(evType, chkType);
+
+        //test channel
+        EXPECT_EQ(event.channel, chn);
+
+        //test data
+        EXPECT_EQ(event.data[0], d0);
+        if(twoData)
+        {
+            EXPECT_EQ(event.data[1], d1);
+        }
+
+        TestEmptyQueue();
+    }
+
+    //help test sysex
+    void TestSysex(uint8_t* msgs, int size)
+    {
+        EXPECT_EQ((uint8_t)event.type, (uint8_t)SystemCommon);
+        EXPECT_EQ((uint8_t)event.sc_type, (uint8_t)SystemExclusive);
+
+        EXPECT_EQ(event.sysex_message_len, size);
+
+        for(int i = 0; i < size; i++)
+        {
+            EXPECT_EQ(event.sysex_data[i], msgs[i]);
+        }
+
+        TestEmptyQueue();
+    }
+
+    //test for empty queue
+    void TestEmptyQueue()
+    {
+        while(midi.HasEvents())
+        {
+            ADD_FAILURE() << "Queue not empty";
+            midi.PopEvent();
+        }
+    }
+
+    MidiHandler<MidiTestTransport> midi;
+    MidiEvent                      event;
+};
+
+//Channel Voice messages
+//for now I'm not using asNoteOff, etc. easier to just test on the raw MidiEvent
+TEST_F(MidiTest, channelVoice)
+{
+    for(uint8_t type = 0; type < 7; type++)
+    {
+        for(uint8_t chn = 0; chn < 16; chn++)
+        {
+            for(uint8_t d0 = 0; d0 < 128; d0++)
+            {
+                for(uint8_t d1 = 0; d1 < 128; d1++)
+                {
+                    uint8_t msgs[]
+                        = {(uint8_t)(0x80 + (type << 4) + chn), d0, d1};
+
+                    bool sendThree = type != 4 && type != 5;
+                    ParseAndPop(msgs, 2 + sendThree);
+
+                    uint8_t chkType
+                        = (type == 3 && d0 > 119) ? (uint8_t)ChannelMode : type;
+                    Test((uint8_t)event.type, chkType, chn, sendThree, d0, d1);
+                }
+            }
+        }
+    }
+}
+
+//Channel Mode Messages
+// also doesn't totally test channels (for now)
+TEST_F(MidiTest, channelMode)
+{
+    //All messages
+    for(uint8_t type = 120; type < 128; type++)
+    {
+        uint8_t msg[] = {0x80 + (3 << 4), type, 0};
+        ParseAndPop(msg, 3);
+        EXPECT_EQ((uint8_t)event.cm_type, type - 120);
+    }
+
+    //LocalControlOn
+    uint8_t msg[] = {0x80 + (3 << 4), 122, 127};
+    ParseAndPop(msg, 3);
+    Test((uint8_t)event.cm_type, (uint8_t)LocalControl, 0, true, 122, 127);
+
+    //MonoModeOn
+    for(uint8_t data = 0; data < 128; data++)
+    {
+        uint8_t msg[] = {0x80 + (3 << 4), 126, data};
+        ParseAndPop(msg, 3);
+        Test((uint8_t)event.cm_type, (uint8_t)MonoModeOn, 0, true, 126, data);
+    }
+}
+
+//System Common Messages (no sysex)
+TEST_F(MidiTest, systemCommon)
```

Comment:
Dito here: I would prefer these as separate tests to also cover the `asXYZ()` functions